include/vara/Utils/Result.h

```diff
@@ -0,0 +1,141 @@
+#ifndef VARA_UTILS_RESULT_H
+#define VARA_UTILS_RESULT_H
+
+#include "vara/Utils/VariantUtil.h"
+
+#include "llvm/Support/raw_ostream.h"
+
+#include <cassert>
+#include <iostream>
+#include <type_traits>
+#include <variant>
+
+namespace result {
+
+template <typename ValueTy>
+class Ok {
+public:
+  Ok(ValueTy V) : V(std::move(V)) {}
+
+  ValueTy operator*() && { return std::move(V); }
+
+  template <
+      typename LValueTy = ValueTy,
+      std::enable_if_t<std::is_copy_constructible_v<LValueTy>, bool> = true>
+  LValueTy operator*() & {
+    return V;
+  }
+
+  template <
+      typename LValueTy = ValueTy,
+      std::enable_if_t<!std::is_copy_constructible_v<LValueTy>, bool> = true>
+  LValueTy operator*() & {
+    return std::move(V);
+  }
+
+  operator bool() const { return true; }
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                       const Ok<ValueTy> &Ok) {
+    OS << Ok.V;
+    return OS;
+  }
+
+private:
+  ValueTy V;
+};
+
+template <>
+class Ok<void> {
+public:
+  Ok() = default;
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                       const Ok<void> &Ok) {
+    OS << "OK";
+    return OS;
+  }
+};
+
+template <typename ErrorTy>
+class Error {
+public:
+  Error(ErrorTy V) : E(std::move(V)) {}
+
+  ErrorTy operator*() && { return std::move(E); }
+
+  template <
+      typename LValueTy = ErrorTy,
+      std::enable_if_t<std::is_copy_constructible_v<LValueTy>, bool> = true>
+  LValueTy operator*() & {
+    return E;
+  }
+
+  template <
+      typename LValueTy = ErrorTy,
+      std::enable_if_t<!std::is_copy_constructible_v<LValueTy>, bool> = true>
+  LValueTy operator*() & {
+    return std::move(E);
+  }
+
+  operator bool() const { return false; }
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                       const Error<ErrorTy> &Error) {
+    OS << Error.E;
+    return OS;
+  }
+
+private:
+  ErrorTy E;
+};
+
+} // namespace result
+
+template <typename ErrorTy, typename ValueTy = void>
+class Result {
+public:
+  Result(result::Ok<ValueTy> V)
+      : Variant(std::forward<result::Ok<ValueTy>>(V)) {}
```

Comment:
What is the idea between forward here?
I guess you miss understood what I meant with my comment. I did not mean to use std::forward here on the Ok value (that would need a move) but to allow the Ok/Error values to construct the carried value by passing the correct parameters.

Reply:
Ah I see, probably we should talk about the remaining points anyway.